### DISCUSSION TITLE
Increasing the layer limit beyond 256 layers

### DISCUSSION THREAD

**Discussion by gorbak25** [containerd/overlaybd#248] [General]
Hi!
Is there a fundamental limitation behind the current 256 layer limit or will overlaybd just work fine after increasing it to 10k? In your paper it's stated the performance should not decrease with increasing layer count, do you have a benchmark backing up that claim?
https://github.com/containerd/overlaybd/blob/2bcbd0b8b4f68e6598705232e0be996d28e92196/src/overlaybd/lsmt/file.cpp#L1599

**Comment by BigVan** [comment]
> will overlaybd just work fine after increasing it to 10k

The maximum number of stacked layers supported by overlaybd is allowed to exceed 256.
We have implemented a nesting mechanism like dolls, which can be found at this link.https://github.com/BigVan/overlaybd/commit/ccadb267a00b15e141aabad3cd9758ce5fde3749
```c++
     /* n + m > 256 */
     auto lsmt_ro_file_0 = open_files_ro( files0, n );
     auto lsmt_ro_file_0 = open_files_ro( files1, m );

     IFile* files[2] = { lsmt_ro_file_0, lsmt_ro_file_1 };
     auto dolls_file = open_files_ro( files, 2 );    
```

> In your paper it's stated the performance should not decrease with increasing layer count, do you have a benchmark backing up that claim

In our paper, we tested hundreds of production images of Alibaba, we found no correlation between the effective number of segments of a merged view image index and the depth of image layers. Therefore the increase in the number of layers does not affect the index lookup.
 You can use 'IMemoryIndex::size()' to checkout the number of segments in your image, like:
```c++
   auto file = stack_files(...);
   auto n_segs = file->index()->size(); 
```

**Comment by lihuiba** [comment]
> Is there a fundamental limitation

No. The limit is only for memory-saving. And the blob format allows arbitrary number of layers.

We use ```unsigned char``` for the layers in current implementation, which is enough in container image scenario. If you need 10k, you can modify the code to use ```uint32_t``` instead.

Index query performance is log(n), where n is the number of segments in merged index.
